src/PostgREST/Plan.hs

```diff
@@ -108,26 +113,93 @@ readPlanTxMode = SQL.Read
 inspectPlanTxMode :: SQL.Mode
 inspectPlanTxMode = SQL.Read
 
+-- | During planning we need to resolve Field -> CoercibleField (finding the context specific target type and map function).
+-- | ResolverContext facilitates this without the need to pass around a laundry list of parameters.
+data ResolverContext = ResolverContext
+  { tables          :: TablesMap
+  , representations :: RepresentationsMap
+  , qi              :: QualifiedIdentifier  -- ^ The table we're currently attending; changes as we recurse into joins etc.
+  , outputType      :: Text                 -- ^ The output type for the response payload; e.g. "csv", "json", "binary".
+  }
+
+resolveColumnField :: Column -> CoercibleField
+resolveColumnField col = CoercibleField (colName col) [] (colNominalType col) Nothing
+
+resolveTableFieldName :: Table -> FieldName -> CoercibleField
+resolveTableFieldName table fieldName =
+  fromMaybe (unknownField fieldName []) $ HMI.lookup fieldName (tableColumns table) >>=
+    Just . resolveColumnField
+
+resolveTableField :: Table -> Field -> CoercibleField
+resolveTableField table (fieldName, []) = resolveTableFieldName table fieldName
+-- If the field is known and a JSON path is given, always assume the JSON type. But don't assume a type for entirely unknown fields.
+resolveTableField table (fieldName, jp) =
+  case resolveTableFieldName table fieldName of
+    tf@CoercibleField{tfIRType=""} -> tf{tfJsonPath=jp}
```

Comment:
The `""` indicates that `CoercibleField` could have a `tfIRType`  of `Maybe Text`. WDYT?

Reply:
Right, glad you brought this up. It's a little tricky so it'll be good to hash it out.

A little background first: the type field actually started as a `Maybe Text`. At that time I was requiring full resolution in all cases, and having a `Maybe` there lead to a technically unreachable panic past the planning stage which got some [push-back in the first round of reviews of this PR](https://github.com/PostgREST/postgrest/pull/2523#discussion_r1003061021). So in #2542, I dropped the `Maybe` and made it impossible to create an 'unresolved' plan.

But resolution is sometimes unnecessary, and critically sometimes not possible: RPC support is out of scope for this PR but we still need it to keep working in the planner!

I experimented with having a sum type like `data CoercibleField = CoercibleField {...} | UnknownField`, but that lead to even more edge cases and an unsightly ballooning up of the code.

So then the choice is between 1. `Maybe`, 2. allow `""`, or 3. both. But it turns out option 1 is option 3.

Bottom line: the reason I ended up with the present solution, making `tfIRType=""` a sentinel for "don't know, don't care", is because there's no easy way to forbid `""` in a compile-time provable manner[1]. So `Maybe` gives us **two** ways to indicate the same thing, which seems redundant and inelegant. In the rare case the type must be known, both `Nothing` and `Just ""` indicate an invariant failure. What do we gain from having two equivalent fault states?

How do you feel given this background?

1: So, obviously you can make a new type such as `NonEmptyText` with a special constructor that forbids the empty string, and not export the default constructor, but that amounts to nothing more than a handshake agreement to not bypass that artificial limitation. Since we're relying on the common sense of future devs, just agreeing on what "" means seems equivalent and leads to less code. (In a dependently typed programming language we could make a compile-time guarantee which would be a great solution here, I think.)